### PR TITLE
Add `input_format` argument to segmentation metrics

### LINKED ISSUE #2563: Error for index tensor verification in GeneralizedDice if target is zeros
## 🐛 Bug

To compute segmentation mask [GeneralizedDice](https://lightning.ai/docs/torchmetrics/stable/segmentation/generalized_dice.html) metrics one can use index tensors as stated by the documentation.

However, if the mask is filled by zeros the conversion to one_hot label will not work in the [following lines](https://github.com/Lightning-AI/torchmetrics/blob/v1.4.0.post0/src/torchmetrics/functional/segmentation/generalized_dice.py#L56C5-L58C1):

```python
if (target.bool() != target).any():  # target is an index tensor
    target = torch.nn.functional.one_hot(target, num_classes=num_classes).movedim(-1, 1)
```

### To Reproduce

Run the following code snippet:

``` python
import torch
from torchmetrics.segmentation import GeneralizedDiceScore

metric = GeneralizedDiceScore(num_classes=3)

target = torch.zeros((2, 3, 3), dtype=torch.int64)
predictions = torch.randint(0, 3, (2, 3, 3), dtype=torch.int64)

metric(predictions, target)

>>> RuntimeError: The size of tensor a (3) must match the size of tensor b (2) at non-singleton dimension 1
```

### Expected behavior

The expected behavior is to have the metric computed after the conversion of the target to a one-hot label.

### Environment

- TorchMetrics version 1.4.0

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #2563
I was foolish to think that the two formats supported in the new segmentation metrics could be differentiated on just based on the value of the tensors. My mistake. To fix this we need a explicit argument, but this opens up for supporting other input formats.

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2572.org.readthedocs.build/en/2572/

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: src/torchmetrics/functional/segmentation/mean_iou.py [hunk 1]
```diff
@@ -33,20 +35,23 @@ def _mean_iou_validate_args(
...
+    input_format: Literal["one-hot", "index"] = "one-hot",
 ) -> Tuple[Tensor, Tensor]:
     """Update the intersection and union counts for the mean IoU computation."""
     _check_same_shape(preds, target)
 
-    if (preds.bool() != preds).any():  # preds is an index tensor
+    if input_format == "index":
         preds = torch.nn.functional.one_hot(preds, num_classes=num_classes).movedim(-1, 1)
-    if (target.bool() != target).any():  # target is an index tensor
+    if input_format == "index":
```
Comment (stancld) [inline src/torchmetrics/functional/segmentation/mean_iou.py:54]: ^^

File: tests/unittests/segmentation/test_generalized_dice_score.py [hunk 2]
```diff
@@ -40,13 +40,14 @@
...
+    input_format: str,
     include_background: bool = True,
     reduce: bool = True,
 ):
     """Calculate reference metric for `MeanIoU`."""
-    if (preds.bool() != preds).any():  # preds is an index tensor
+    if input_format == "index":
         preds = torch.nn.functional.one_hot(preds, num_classes=NUM_CLASSES).movedim(-1, 1)
-    if (target.bool() != target).any():  # target is an index tensor
+    if input_format == "index":
```
Comment (stancld) [inline tests/unittests/segmentation/test_generalized_dice_score.py:50]: ^^

File: tests/unittests/segmentation/test_mean_iou.py [hunk 3]
```diff
@@ -40,14 +40,15 @@
...
     include_background: bool = True,
     per_class: bool = True,
     reduce: bool = True,
 ):
     """Calculate reference metric for `MeanIoU`."""
-    if (preds.bool() != preds).any():  # preds is an index tensor
+    if input_format == "index":
         preds = torch.nn.functional.one_hot(preds, num_classes=NUM_CLASSES).movedim(-1, 1)
-    if (target.bool() != target).any():  # target is an index tensor
+    if input_format == "index":
```
Comment (stancld) [inline tests/unittests/segmentation/test_mean_iou.py:51]: ^^

**Review by stancld — APPROVED** [review]

File: src/torchmetrics/functional/segmentation/generalized_dice.py [hunk 4]
```diff
@@ -45,15 +48,16 @@ def _generalized_dice_update(
...
     """Update the state with the current prediction and target."""
     _check_same_shape(preds, target)
     if preds.ndim < 3:
         raise ValueError(f"Expected both `preds` and `target` to have at least 3 dimensions, but got {preds.ndim}.")
 
-    if (preds.bool() != preds).any():  # preds is an index tensor
+    if input_format == "index":
         preds = torch.nn.functional.one_hot(preds, num_classes=num_classes).movedim(-1, 1)
-    if (target.bool() != target).any():  # target is an index tensor
+    if input_format == "index":
```
[the remark this answers is not in the document]
Comment (Borda) [reply src/torchmetrics/functional/segmentation/generalized_dice.py:60]:
```suggestion
```

File: src/torchmetrics/functional/segmentation/mean_iou.py (hunk 1 above)
Comment (Borda) [reply src/torchmetrics/functional/segmentation/mean_iou.py:54]:
```suggestion
```

File: tests/unittests/segmentation/test_generalized_dice_score.py (hunk 2 above)
Comment (Borda) [reply tests/unittests/segmentation/test_generalized_dice_score.py:50]:
```suggestion
```

File: tests/unittests/segmentation/test_mean_iou.py (hunk 3 above)
Comment (Borda) [reply tests/unittests/segmentation/test_mean_iou.py:51]:
```suggestion
```

**Comment by Borda** [comment]
ok, I missed it for `1.4.2` but I believe can have `1.5.0` shortly

**Pull request merged**